src/include/cell.i

```diff
@@ -209,72 +226,86 @@ __wt_cell_pack_data(WT_CELL *cell, uint64_t rle, size_t size)
+		}
+		(void)__wt_vpack_uint(&p, 0, (uint64_t)size);	/* Length */
 	}
-	(void)__wt_vpack_uint(&p, 0, (uint64_t)size);	/* Length */
 	return (WT_PTRDIFF(p, cell));
 }
 
 /*
  * __wt_cell_pack_data_match --
- *	Return if two items would have identical WT_CELLs (except for any RLE).
+ *	Return if two items would have identical WT_CELLs (except for timestamps
+ * and any RLE).
  */
 static inline int
-__wt_cell_pack_data_match(
-    WT_CELL *page_cell, WT_CELL *val_cell, const uint8_t *val_data,
-    bool *matchp)
+__wt_cell_pack_data_match(WT_CELL *page_cell,
+    WT_CELL *val_cell, const uint8_t *val_data, bool *matchp)
 {
-	uint64_t av, bv;
+	uint64_t alen, blen, v;
 	const uint8_t *a, *b;
 	bool rle;
 
-	*matchp = 0;				/* Default to no-match */
+	*matchp = false;			/* Default to no-match */
 
 	/*
 	 * This is a special-purpose function used by reconciliation to support
 	 * dictionary lookups.  We're passed an on-page cell and a created cell
 	 * plus a chunk of data we're about to write on the page, and we return
 	 * if they would match on the page.  The column-store comparison ignores
-	 * the RLE because the copied cell will have its own RLE.
+	 * the timestamps and the RLE because the copied cell will have its own.
 	 */
 	a = (uint8_t *)page_cell;
 	b = (uint8_t *)val_cell;
 
 	if (WT_CELL_SHORT_TYPE(a[0]) == WT_CELL_VALUE_SHORT) {
-		av = a[0] >> WT_CELL_SHORT_SHIFT;
+		alen = a[0] >> WT_CELL_SHORT_SHIFT;
 		++a;
+		if (__wt_process.page_version_ts) {
+			WT_RET(__wt_vunpack_uint(&a, 0, &v));	/* Skip TS */
+			WT_RET(__wt_vunpack_uint(&a, 0, &v));
+		}
 	} else if (WT_CELL_TYPE(a[0]) == WT_CELL_VALUE) {
-		rle = (a[0] & WT_CELL_64V) != 0;	/* Skip any RLE */
+		rle = (a[0] & WT_CELL_64V) != 0;
 		++a;
-		if (rle)
-			WT_RET(__wt_vunpack_uint(&a, 0, &av));
-		WT_RET(__wt_vunpack_uint(&a, 0, &av));	/* Length */
+		if (__wt_process.page_version_ts) {
+			WT_RET(__wt_vunpack_uint(&a, 0, &v));	/* Skip TS */
+			WT_RET(__wt_vunpack_uint(&a, 0, &v));
+		}
+		if (rle)					/* Skip RLE */
+			WT_RET(__wt_vunpack_uint(&a, 0, &v));
+		WT_RET(__wt_vunpack_uint(&a, 0, &alen));	/* Length */
 	} else
 		return (0);
 
 	if (WT_CELL_SHORT_TYPE(b[0]) == WT_CELL_VALUE_SHORT) {
-		bv = b[0] >> WT_CELL_SHORT_SHIFT;
+		blen = b[0] >> WT_CELL_SHORT_SHIFT;
 		++b;
+		if (__wt_process.page_version_ts) {
+			WT_RET(__wt_vunpack_uint(&b, 0, &v));	/* Skip TS */
+			WT_RET(__wt_vunpack_uint(&b, 0, &v));
+		}
 	} else if (WT_CELL_TYPE(b[0]) == WT_CELL_VALUE) {
-		rle = (b[0] & WT_CELL_64V) != 0;	/* Skip any RLE */
+		rle = (b[0] & WT_CELL_64V) != 0;
 		++b;
-		if (rle)
-			WT_RET(__wt_vunpack_uint(&b, 0, &bv));
-		WT_RET(__wt_vunpack_uint(&b, 0, &bv));	/* Length */
+		if (__wt_process.page_version_ts) {
+			WT_RET(__wt_vunpack_uint(&b, 0, &v));	/* Skip TS */
+			WT_RET(__wt_vunpack_uint(&b, 0, &v));
+		}
+		if (rle)					/* Skip RLE */
+			WT_RET(__wt_vunpack_uint(&b, 0, &v));
+		WT_RET(__wt_vunpack_uint(&b, 0, &blen));	/* Length */
```

Comment:
The repetition in this code feels a bit unwieldy, but it's performance sensitive so possibly not worth refactoring.

Reply:
I'm open to refactoring this, can you share further the path you're seeing?

Reply:
The code from lines 268->286 and 288->306 are the same apart from operating on `a` or `b` - that code could be pulled into a shared function.

It'd also be possible to make an inline function called `__skip_encoded_timestamps` (or similar) that does:
```
 		if (__wt_process.page_version_ts) {
 			WT_RET(__wt_vunpack_uint(&b, 0, &v));	/* Skip TS */
 			WT_RET(__wt_vunpack_uint(&b, 0, &v));
 		}
```

Reply:
I've futzed with this a little, and I don't see anything that I like better than the current code. Unless you feel strongly, I'm planning to leave this alone.